File: benchmarks/micro/StackNestedFieldForwarding.c
#include <stdint.h>
#include <unistd.h>

typedef struct Inner {
    int64_t value;
    int64_t salt;
} Inner;

typedef struct Outer {
    Inner left;
    Inner right;
} Outer;

static int64_t accumulate(int64_t seed, int64_t salt) {
    Outer outer = {
        { seed, salt },
        { salt, seed }
    };
    outer.left.value = outer.left.value + outer.right.salt;
    int64_t first = outer.left.value;
    outer.right.value = first ^ (outer.right.value >> 5);
    return outer.left.value + outer.right.value + outer.left.salt;
}

int main(void) {
    int64_t total = (int64_t)getpid();

    for (int32_t i = 0; i < 200000; i += 1) {
        total += accumulate(total % 1000000007, (int64_t)i);
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/StackNestedFieldForwarding.rs
struct Inner {
    value: i64,
    salt: i64,
}

struct Outer {
    left: Inner,
    right: Inner,
}

fn accumulate(seed: i64, salt: i64) -> i64 {
    let mut outer = Outer {
        left: Inner { value: seed, salt },
        right: Inner {
            value: salt,
            salt: seed,
        },
    };
    outer.left.value = outer.left.value + outer.right.salt;
    let first = outer.left.value;
    outer.right.value = first ^ (outer.right.value >> 5);
    outer.left.value + outer.right.value + outer.left.salt
}

fn main() {
    let mut total = std::process::id() as i64;

    for i in 0_i32..200_000 {
        total += accumulate(total % 1_000_000_007, i as i64);
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
